engine/table/src/main/java/io/deephaven/engine/table/impl/hierarchical/TreeTableImpl.java

```diff
@@ -0,0 +1,422 @@
+
+    private static final ColumnName TREE_COLUMN = ColumnName.of("__TREE__");
+
+    private final ColumnSource<?> sourceParentIdSource;
+    private final QueryTable tree;
+    private final AggregationRowLookup treeRowLookup;
+
+    private final ColumnSource<Table> treeNodeTableSource;
+    private final TreeSourceRowLookup sourceRowLookup;
+    private final boolean filtered;
+    private final ColumnName identifierColumn;
+    private final ColumnName parentIdentifierColumn;
+    private final Set<ColumnName> nodeFilterColumns;
+    private final TreeNodeOperationsRecorder nodeOperations;
+    private final TableDefinition snapshotDefinition;
+
+    private TreeTableImpl(
+            @NotNull final Map<String, Object> initialAttributes,
+            @NotNull final QueryTable source,
+            @NotNull final QueryTable tree,
+            @NotNull final TreeSourceRowLookup sourceRowLookup,
+            @NotNull final ColumnName identifierColumn,
+            @NotNull final ColumnName parentIdentifierColumn,
+            @NotNull final Set<ColumnName> nodeFilterColumns,
+            @Nullable final TreeNodeOperationsRecorder nodeOperations,
+            @Nullable final TableDefinition snapshotDefinition) {
+        super(initialAttributes, source, getTreeRoot(tree));
+        if (source.isRefreshing()) {
+            manage(tree);
+            manage(sourceRowLookup);
+        }
+        sourceParentIdSource = source.getColumnSource(parentIdentifierColumn.name());
+        this.tree = tree;
+        treeRowLookup = AggregationProcessor.getRowLookup(tree);
+        treeNodeTableSource = tree.getColumnSource(TREE_COLUMN.name(), Table.class);
+        this.sourceRowLookup = sourceRowLookup;
+        filtered = !sourceRowLookup.sameSource(source);
+        this.identifierColumn = identifierColumn;
+        this.parentIdentifierColumn = parentIdentifierColumn;
+        this.nodeFilterColumns = nodeFilterColumns;
+        this.nodeOperations = nodeOperations;
+        this.snapshotDefinition = snapshotDefinition == null
+                ? computeSnapshotDefinition(getNodeDefinition())
+                : snapshotDefinition;
+
+        if (source.isRefreshing()) {
+            Assert.assertion(tree.isRefreshing(), "tree.isRefreshing() if source.isRefreshing()");
+            // The tree aggregation result depends on the source and all the node tables.
+            manage(tree);
+            // The reverse lookup just depends on the (original, unfiltered) source, which may not be our direct source.
+            manage(sourceRowLookup);
+        }
+    }
+
+    @Override
+    public String getDescription() {
+        return "TreeTable(" + source.getDescription()
+                + ", " + identifierColumn.name()
+                + ", " + parentIdentifierColumn.name()
+                + ")";
+    }
+
+    @Override
+    public Table getRootOnlyKeyTable() {
+        return makeNullIdentifierColumnSingleRowTable(getSource(), getIdentifierColumn());
```

Comment:
The implementations pre-date when I reached the conclusion that I needed to always expand the root by default, rather than requiring the user to send it to me. I kept it around because I thought it might be a useful way to communicate the root node key to clients. I could change it to an empty table with the correct schema.

This is used for bootstrapping; the client doesn't know about any keys to expand until it snapshots a view that includes those keys, so we let them create a view without specifying expansions and instead give them a default with only the root expanded.